### DISCUSSION TITLE
Airflow 2.2.4: schedular going down for 1-2 minute on every 10 minute

### DISCUSSION THREAD

**Discussion by dviru** [apache/airflow#22563] [Q&A]
Hi Team, I am using airflow 2.2.4 and deployed it on aws eks cluster. I noticed that every 5-10 minute schedular down message I am seeing on airflow UI. When I checked airflow schedular log I am seeing the lot of below statements I am seeing  

`[2022-03-21 08:21:21,640] {kubernetes_executor.py:729} INFO - Attempting to adopt pod sampletask.05b6f567b4a64bd5beb16e526ba94d7a` 

This above statement will print for all completed pod which exist in eks, But it is repeating multiple time and as also invoking the PATCH api. But I have one question. Why below code pulling all the completed pod details for every time form EKS and updating. Once in it patch the completed pod. Next time it should pull that updated POD details. What happing this activity for 1000 it is finishing 1 minute, for 7000 pod it taking 3-5 minute thats the reason scheduler is going down. Please let me if my understanding is wrong. 

<img width="1054" alt="Screenshot 2022-03-21 at 8 53 03 PM" src="https://user-images.githubusercontent.com/10843400/160352813-9ff57de3-782f-4cee-8f7c-f6d5b8a60d29.png">

As of now we are not deleting the completed pod at least for 3 days to debug something.

**Comment by potiuk** [comment]
I think you simply have not enough resuources (memory?)  - look at your k8s logs and check why scheduler gets killed

**Reply by dviru** [reply]
@potiuk I checked we are using "m5.xlarge" it has 16GB memory and 4 vCPUs. This instance is only dedicated to schedular.

**Reply by dviru** [reply]
@potiuk looks like the above "**adopt_completed_pods**" method need call only when we set "**delete_worker_pods**" to true. other wise pull all pods from EKS and calling PATCH api for every POD it is time consuming. In my case we have 6000-7000 completed pods in EKS cluster.
Sorry if my understanding is wrong.

**Reply by potiuk** [reply]
I do not know all the details there I think there are few other people (@dstandish @ashb @uranusjr @ephraimbuddy ) who know a bit more about the internal of K8S executor more than I do. I think there were some improvements recently in this area (not sure if they are parts of the upcoming 2.3.0 or 2.2.5 which is just being voted) but I do recall similar issues when a number of PODs caused similar issues.

**Reply by uranusjr** [reply]
I don’t have much to offer for this one; the function has been like this since the beginning, it’s difficult to tell the reasoning behind it, other than “it works” and “it is simple”.

cc @dimberman (original author) for context.

**Comment by dviru** [comment]
@dstandish @ashb @uranusjr @ephraimbuddy Any update on this. Is I am doing any thing wrong or it is issue in airflow 2.2.4?

**Reply by potiuk** [reply]
Please be mindful for people time @dviru - people here are in different timzeones and responding in their free time. Sometimes it will be days until you get answer. Be patient please.

**Reply by potiuk** [reply]
or even weeks.

**Reply by dviru** [reply]
sure @potiuk. Sorry for that.

**Reply by potiuk** [reply]
I  discussed it with the "wise ones" - and the thing seems to be related to ``delete_worker_pods = False`` has never been meant to be used outside of development/debugging context. So if you are running it now in PROD, you **might** get this kind of problems. For now I recommend you to disable it and only set ``delete_worker_pods_on_failure = False`` and ``delete_worker_pods = True``. This will leave the failing pods for investigation but delete all the "succesful" ones. 

In the meantime - can you please open an issue about it - this seems to be possible to be implemented in 2.3.0 to make those queries more performant in case you also leave the Pods, but this is likely not something you can fix in the current versions easly.

**Reply by jedcunningham** [reply]
I'm curious why you'd want to keep all pods in prod? +1 on opening an issue.

**Reply by dviru** [reply]
Sure @potiuk as of now we will set `delete_worker_pods = True`.
@jedcunningham regarding your question. In our cluster some the DAGs are using the K8s operator. Some time users need to see the logs of the PODs while debugging the issue. thats the reason at least we are keeping 2-3 days completed DAGs.

**Reply by potiuk** [reply]
Might be a good feature actually @dimberman @jedcunningham to add retention time for PODs.

**Reply by potiuk** [reply]
That would remove the need of having "handled" label at the expense of doing a bit more elborated housekeeping query.

**Reply by jedcunningham** [reply]
In general I think having pod logs aggregated somewhere is a better solution than relying on the actual pods to stick around (e.g. getting logs into elasticsearch or equivalent). This puts them in a more useful location and can help with discovering trends that could be missed looking pod by pod over a short window. You then also aren't reliant on a node sticking around either.

Our `airflow kubernetes cleanup-pods` cli could probably stand a new flag to consider age as well, but I'm not sure getting the scheduler involved makes sense in general.

I can see the need for an equivalent of `delete_worker_pods_on_failure` for KPO though (in fact, @eladkal brought it up today go figure).

**Reply by dviru** [reply]
I agree with you @jedcunningham. Please suggest me shall raise this as issue or not?

**Reply by jedcunningham** [reply]
Yes please! At a bare minimum, we need to add a warning to the docs for `delete_worker_pods`, and possibly/more likely handle them sticking around more gracefully.

**Reply by dviru** [reply]
Sure I will raise the issue. Thanks @jedcunningham and @potiuk for your support.

**Reply by dviru** [reply]
Raised the issue: #22612

**Reply by eladkal** [reply]
@jedcunningham its not just logs.
With the issue we had today logs were empty the clue to figure out the problem was in the pod events.
